src/fromager/dependency_graph.py

```diff
@@ -279,12 +280,16 @@ def get_install_dependency_versions(
     def _depth_first_traversal(
         self,
         start_node: list[DependencyEdge],
+        visited: set[str],
         match_dep_types: list[RequirementType] | None = None,
     ) -> typing.Iterable[DependencyEdge]:
         for edge in start_node:
+            if edge.destination_node.key in visited:
+                continue
             if match_dep_types and edge.req_type not in match_dep_types:
                 continue
+            visited.add(edge.destination_node.key)
```

Comment:
We should add the start nodes, too, right?

Reply:
they will get added when they are being process in the loop

Reply:
I don't think the start nodes from `self.nodes[ROOT].children` on line 243 will be added unless they also appear as a destination node of something else, right?

One way to address it would be to add `start_node` to the `visited` set as the first thing this function does. That way every node that is traversed is added. I think if you make that change, you don't need line 291, because the set will be updated during the recursion on line 293.

Reply:
Ah, OK.